### DISCUSSION TITLE
Mellinger control.

### DISCUSSION THREAD

**Discussion by Carmol1** [bitcraze/discussions#189] [Q&A] [answered]
Hello.

By observing  the done in the paper 'Minimum snap trajectory generation and control for quadrotors' is easier to comprehend the controller_mellinger.c, however, the doubt arises in the code lines 168, 177, 185 and 189, since, I do not understand why are they used?, as far as I know modeAbs is eqaul to 1 and modeVelocity is equal to 2, but setpoint->mode.x what does values take?.

I'd appreciate your help.

**Comment by krichardsson** [comment, marked as answer]
The [setpoint_t](https://github.com/bitcraze/crazyflie-firmware/blob/8c606d1f8f89f1f00462dfa1f294a431a321242d/src/modules/interface/stabilizer_types.h#L190-L211) structure is very flexible and is (sort of) [documented on a CRTP level](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/crtp/crtp_generic_setpoint/) 

It is possible to set [the mode](https://github.com/bitcraze/crazyflie-firmware/blob/8c606d1f8f89f1f00462dfa1f294a431a321242d/src/modules/interface/stabilizer_types.h#L184-L188) for each component, so for instance it would be possible to use a setpoint with an absolute value for x (say 2.0 m) while you specify a velocity for y (1 m/s). 

Note that all controllers might not be able to handle all possible permutations, but more likely the "standard" use cases such as flying manually or absolution positions.

What you see in the code is the handling of different modes.

**Reply by Carmol1** [reply]
Thank you for your response.

But, in the instance is fulfilled 

else {
    target_thrust.x = -sinf(radians(setpoint->attitude.pitch));
    target_thrust.y = -sinf(radians(setpoint->attitude.roll));
    // In case of a timeout, the commander tries to level, ie. x/y are disabled, but z will use the previous setting
    // In that case we ignore the last feedforward term for acceleration
    if (setpoint->mode.z == modeAbs) {
      target_thrust.z = g_vehicleMass * GRAVITY_MAGNITUDE + kp_z  * r_error.z + kd_z  * v_error.z + ki_z  * i_error_z;
    } else {
      target_thrust.z = 1;
    }

and, so I understand the line "if (setpoint->mode.x == modeAbs)"  is only fulfilled if I use x=1 or not?.

I'd appreciate your help.

**Reply by krichardsson** [reply]
I'm not sure I fully understand your question,  but I'll try to elaborate a bit.

>  so I understand the line "if (setpoint->mode.x == modeAbs)" is only fulfilled if I use x=1 or not?.

It depends on which x you are referring to. The `setpoint->mode.x` tells the controller how to interpret the rest of the data (related to x) in the setpoint structure. I think it could be a good idea to look at the crtp_commander_generic.c file to get a feel of how the setpoint structure is used, this is where incoming setpoint packets are converted to the setpoint structure. One example could be the [Z-distance style of setpoint for instance](https://github.com/bitcraze/crazyflie-firmware/blob/8c606d1f8f89f1f00462dfa1f294a431a321242d/src/modules/src/crtp_commander_generic.c#L126-L148). This setpoint type is used for manual flight with a game pad where the user is flying around at a fixed height. The game pad is mapped to roll/pitch/yaw as usual, but instead of controlling the thrust, the user controls the height. As you can see in the code, the mode for Z is set to `modeAbs` and the position for Z is set to the desired height.

``` C
setpoint->mode.z = modeAbs;
setpoint->position.z = values->zDistance;
```

One thing to note is that there seems to be some assumptions in the code, for instance that x and y have the same mode, even if it is not checked. One example is [this code](https://github.com/bitcraze/crazyflie-firmware/blob/8c606d1f8f89f1f00462dfa1f294a431a321242d/src/modules/src/controller_mellinger.c#L168-L182) where only the mode of x is checked while the same logic is applied to y as well.

The Mellinger controller was kindly contributed by @whoenig and @jpreiss , do you want to add anything here?

**Reply by whoenig** [reply]
I think you covered it pretty well. In the original paper, the controller is only used for trajectory tracking and has therefore a position and an attitude controller. The mode switches are introduced to allow flying manually with a joystick as well (by essentially disabling the position controller part). The main use-case is for easier tuning of Bolt-enabled drones.

**Reply by Carmol1** [reply]
Thank you so much for your help, by looking at crtp_commander_generic.c, I have already understood his explanation, once again I appreciate your help.
